src/providers/mod.rs

```diff
@@ -0,0 +1,65 @@
+use failure::Error;
+use std::collections::HashMap;
+
+/// This is used to group machines into connections
+/// to cloud providers. e.g., for AWS we need a separate
+/// connection to each region.
+pub trait MachineSetup {
+    type Region: Eq + std::hash::Hash + Clone + std::string::ToString;
+    fn region(&self) -> Self::Region;
+}
+
+/// Implement this trait to implement a new cloud provider for Tsunami.
+/// Tsunami will call `init_instances` once per unique region, as defined by `MachineSetup`.
+pub trait Launcher: Drop + Send + Sync + Sized {
+    type Region: Send + Eq + std::hash::Hash + Clone + std::string::ToString;
+    type Machine: MachineSetup<Region = Self::Region> + Send;
+
+    fn init(log: slog::Logger, r: Self::Region) -> Result<Self, Error>;
```

Comment:
This kind of pattern is pretty weird. I would instead expect that the trait `Launcher` had a method like:
```rust
struct LaunchDescription {
        r: Self::Region
        max_instance_duration: Option<std::time::Duration>,
        max_wait: Option<std::time::Duration>,
        machines: impl IntoIterator<Item = (String, Self::Machine)>,
}

trait Launcher {
    type Machine: Instance;
    fn launch(
        &mut self,
        desc: LaunchDescription
    ) -> HashMap<String, Self::Machine>;
}

trait Instance {
    fn connect(&self) -> Result<crate::Machine, Error>;
}
```
and then that `TsunamiBuilder::spawn` _takes_ an `l: Launcher`.

Reply:
One nice thing about this API is that it auto-groups the `descriptors` into `Self::Region`s, and only initializes the correct number of `impl Launcher`s (see https://github.com/akshayknarayan/tsunami/blob/machine-lifetimes/src/lib.rs#L173).
Passing in an `l: Launcher` would preclude that, and all the callers would have to implement de-duplication like this, which is a little annoying. This is also why `type Region: Hash` is there.

I do think a marker trait like you suggested above might be useful, but these two questions should probably be resolved together.

Reply:
One nice thing about this API is that it auto-groups the `descriptors` into `Self::Region`s, and only initializes the correct number of `impl Launcher`s (see https://github.com/akshayknarayan/tsunami/blob/machine-lifetimes/src/lib.rs#L173).
Passing in an `l: Launcher` would preclude that, and all the callers would have to implement de-duplication like this, which is a little annoying. This is also why `type Region: Hash` is there.

I do think a marker trait like you suggested above might be useful, but these two questions should probably be resolved together.

Reply:
You can still group them like that inside of `spawn`, since it still knows about each region from the `LaunchDescription`. So if you want them grouped, you can get them grouped.